new-new-packages/compiler/src/compiler/string_to_rcst.rs

```diff
@@ -0,0 +1,1394 @@
+        ))
+    }
+    #[test]
+    fn test_lambda() {
+        assert_eq!(lambda("2", 0), None);
+        assert_eq!(
+            lambda("{ 2 }", 0),
+            Some((
+                "",
+                Rcst::Lambda {
+                    opening_curly_brace: Box::new(Rcst::OpeningCurlyBrace),
+                    parameters_and_arrow: None,
+                    body: vec![Rcst::Int(2)],
+                    closing_curly_brace: Box::new(Rcst::ClosingCurlyBrace),
+                }
+            ))
+        );
+        // { a ->
+        //   foo
+        // }
+        assert_eq!(
+            lambda("{ a ->\n  foo\n}", 0),
+            Some((
+                "",
+                Rcst::Lambda {
+                    opening_curly_brace: Box::new(Rcst::OpeningCurlyBrace),
+                    parameters_and_arrow: Some((
+                        vec![Rcst::Identifier("a".to_string())],
+                        Box::new(Rcst::Arrow)
+                    )),
+                    body: vec![Rcst::Identifier("foo".to_string())],
+                    closing_curly_brace: Box::new(Rcst::ClosingCurlyBrace),
+                }
+            ))
+        );
+        // {
+        // foo
+        assert_eq!(
+            lambda("{\nfoo", 0),
+            Some((
+                "\nfoo",
+                Rcst::Lambda {
+                    opening_curly_brace: Box::new(Rcst::OpeningCurlyBrace),
+                    parameters_and_arrow: None,
+                    body: vec![],
+                    closing_curly_brace: Box::new(Rcst::Error {
+                        unparsable_input: "".to_string(),
+                        error: RcstError::CurlyBraceNotClosed
+                    }),
+                }
+            ))
+        );
+        // {->
+        // }
+        assert_eq!(
+            lambda("{->\n}", 1),
+            Some((
+                "\n}",
+                Rcst::Lambda {
+                    opening_curly_brace: Box::new(Rcst::OpeningCurlyBrace),
+                    parameters_and_arrow: Some((vec![], Box::new(Rcst::Arrow))),
+                    body: vec![],
+                    closing_curly_brace: Box::new(Rcst::Error {
+                        unparsable_input: "".to_string(),
+                        error: RcstError::CurlyBraceNotClosed
+                    }),
+                }
+            ))
+        );
+    }
+
+    fn assignment(input: &str, indentation: usize) -> Option<(&str, Rcst)> {
+        log::trace!("assignment({:?}, {:?})", input, indentation);
+        let (input, mut signature) = run_of_expressions(input, indentation)?;
+        if signature.is_empty() {
+            return None;
+        }
+
+        let (input, whitespace) = whitespaces_and_newlines(input, indentation + 1, true);
+        let last = signature.pop().unwrap();
+        signature.push(last.wrap_in_whitespace(whitespace.clone()));
+
+        let parameters = signature.split_off(1);
+        let name = signature.into_iter().next().unwrap();
```

Comment:
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.first

Reply:
We need the `Rcst`, but there's only `.first()` returning `&Rcst` and `.first_mut()` returning `&mut Rcst`. There would also be a nice `.split_first()` method on `[T]` to do `let (name, parameters) = signature.split_first();`, but it has the same problem.